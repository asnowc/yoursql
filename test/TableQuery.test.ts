import { DbTableQuery, pgSqlTransformer, SqlValuesCreator } from "@asnc/yoursql";
import { test, expect, describe } from "vitest";

describe("TableQuery", function () {
  interface Table {
    name: string;
    level: number;
    id: bigint;
  }
  interface CreateTable {
    name: string;
    level?: number;
  }
  const table = new DbTableQuery<Table, CreateTable>("user", new SqlValuesCreator(pgSqlTransformer));
  describe("select", function () {
    test("select-columns", function () {
      expect(table.fromAs().select("*").toString()).toMatchSnapshot();
      expect(table.fromAs().select({ name: true, rename_level: "level", id: "id" }).toString()).toMatchSnapshot();
    });
    test("select-option", function () {
      const sql = table
        .fromAs()
        .select("*")
        .where("limit IS NULL")
        .orderBy(["id ASC NULLS FIRST"])
        .limit(100, 23)
        .toString();
      expect(sql).toMatchSnapshot();
    });
  });
  describe("update", function () {
    test("update-base", function () {
      expect(table.update({ id: 78n })).toMatchSnapshot();
      expect(table.update({ id: 78n }, { where: "id IS NULL" })).toMatchSnapshot();
    });
    test("update-error", function () {
      expect(() => table.update({}), "值为空").toThrowError();
    });
    test("updateWithResult", function () {
      expect(
        table.updateWithResult({ id: 78n }, { level: "rename", name: true }).toString(),
        "返回结果重命名"
      ).toMatchSnapshot();

      expect(() => table.updateWithResult({ id: 78n }, {})).toThrowError();
    });
  });
  describe("delete", function () {
    test("delete", function () {
      expect(table.delete({ where: "id IS NULL" })).toMatchSnapshot();
    });
    test("deleteWithResult", function () {
      expect(table.deleteWithResult("*").toString()).toMatchSnapshot();
      expect(table.deleteWithResult({ level: "rename", name: true }).toString(), "返回结果重命名").toMatchSnapshot();

      expect(() => table.deleteWithResult({})).toThrowError();
    });
  });
  describe("insert", function () {
    test("insert", function () {
      expect(
        table.insert([
          { name: "张三", level: 8 },
          { name: "王五", level: 9 },
        ])
      ).toMatchSnapshot();

      expect(
        table.insert([{ name: "张三" }], { conflict: ["id", "level"], updateValues: { level: 89 } }),
        "冲突更新"
      ).toMatchSnapshot();

      expect(table.insert("VALUES('张三')", ["name"]), "字符串值").toMatchSnapshot();
    });
    test("insert-error", function () {
      expect(() => table.insert([])).toThrowError();
      expect(() => table.insert([{}, {}] as any)).toThrowError();
      expect(() => table.insert({} as any)).toThrowError();
      //@ts-ignore
      expect(() => table.insert("VALUES(1)"), "没有指定列").toThrowError();
      expect(() => table.insert("VALUES(1)", []), "没有指定列").toThrowError();
    });
    test("insertWithResult", function () {
      const values = [{ name: "hh" }];
      expect(
        table.insertWithResult(values, { level: "rename", name: true }).toString(),
        "返回结果重命名"
      ).toMatchSnapshot();

      expect(() => table.insertWithResult([], { level: true }), "插入值不能为空").toThrowError();
      expect(() => table.insertWithResult(values, {}), "返回结果不能为空").toThrowError();
    });
  });
});
