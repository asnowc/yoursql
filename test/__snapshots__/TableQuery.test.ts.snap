// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TableQuery > delete > delete 1`] = `
"DELETE FROM user
WHERE id IS NULL"
`;

exports[`TableQuery > delete > deleteWithResult 1`] = `
"DELETE FROM user
RETURNING *"
`;

exports[`TableQuery > delete > deleteWithResult 2`] = `
"DELETE FROM user
RETURNING level AS rename, name"
`;

exports[`TableQuery > insert > insert 1`] = `
"INSERT INTO user (name,level)
VALUES
('张三',8),
('王五',9)"
`;

exports[`TableQuery > insert > insert 2`] = `
"INSERT INTO user (name)
VALUES
('张三')
ON CONFLICT (id,level)
DO UPDATE SET
level = 89"
`;

exports[`TableQuery > insert > insert 3`] = `
"INSERT INTO user (name)
VALUES('张三')"
`;

exports[`TableQuery > insert > insertWithResult 1`] = `
"INSERT INTO user (name)
VALUES
('hh')
RETURNING level AS rename, name"
`;

exports[`TableQuery > select > select-columns 1`] = `
"SELECT *
FROM user"
`;

exports[`TableQuery > select > select-columns 2`] = `
"SELECT name,level AS rename_level,id AS id
FROM user"
`;

exports[`TableQuery > select > select-option 1`] = `
"SELECT *
FROM user
WHERE limit IS NULL
ORDER BY id ASC NULLS FIRST
LIMIT 100
OFFSET 23"
`;

exports[`TableQuery > update > update-base 1`] = `
"UPDATE user
SET id = 78"
`;

exports[`TableQuery > update > update-base 2`] = `
"UPDATE user
SET id = 78
WHERE id IS NULL"
`;

exports[`TableQuery > update > updateWithResult 1`] = `
"UPDATE user
SET id = 78
RETURNING level AS rename, name"
`;
